### ISSUE TITLE
code stuck at All DDP processes registered

### ISSUE THREAD

**Issue by derrick-xwp** [PyTorchLightning/pytorch-lightning#9641]
initializing ddp: GLOBAL_RANK: 1, MEMBER: 2/2
----------------------------------------------------------------------------------------------------
distributed_backend=nccl
All DDP processes registered. Starting ddp with 2 processes
----------------------------------------------------------------------------------------------------

**Comment by tchaton** [comment]
Dear @derrick-xwp,

Would you mind providing more content for us to understand the problem and to be able to help.
The issue you opened would be closed otherwise.

Best,
T.C

**Comment by awaelchli** [comment]
You can run your script with `NCCL_DEBUG=WARN` to see if there is anything wrong with the nccl communication. Can you also verify gloo works?

**Comment by lifefeel** [comment]
I had the same problem. 

Using:
- NCCL version 2.11.4+cuda11.6
- pytorch-lightning 1.5.10

It works after setting `PL_TORCH_DISTRIBUTED_BACKEND=gloo`.

**Comment by Vadkoz** [comment]
Same issue.
And yeah, it works after setting PL_TORCH_DISTRIBUTED_BACKEND=gloo.

But gloo and 2 GPUs works slower than 1 single gpu.

**Comment by andreemic** [comment]
Same issue... NCCL_DEBUG=WARN doesn't gvie ant Stuck forever after:

```
...
initializing ddp: GLOBAL_RANK: 3, MEMBER: 4/4
----------------------------------------------------------------------------------------------------
distributed_backend=nccl
All DDP processes registered. Starting ddp with 4 processes
----------------------------------------------------------------------------------------------------

NCCL version 2.14.3+cuda11.8

```

I'm using a PyTorch dataloader wrapped around a Hugging Face dataset.

**Comment by kingwmk** [comment]
> Closing due to inactivity. Please check other related issues here on Github where we have provided help for hanging processes.
> 
> Possible reasons:
> 
> * not the latest PL version, some bugs were there in older PL versions. Solution: update to latest
> * you are running with ddp_spawn and many workers for the dataloader. it's possible that its not stuck and you just need to wait for a while. Solution: switch to `Trainer(accelerator="ddp")`
> * distributed_backend=nccl not compatible or incorrect pytorch version installed (try a clean environment with fresh and latest pytorch installed), or try setting `PL_TORCH_DISTRIBUTED_BACKEND=gloo`

Hello. May I ask How to setting `PL_TORCH_DISTRIBUTED_BACKEND=gloo`? Could you give an exmple command line?
Thank you!!!!

**Comment by nijisakai** [comment]
> PL_TORCH_DISTRIBUTED_BACKEND=gloo

```
import os
os.environ["PL_TORCH_DISTRIBUTED_BACKEND"] = "gloo"
```
but not work for me

**Comment by awaelchli** [comment]
In the latest versions of Lightning, you set it like this:

```py
# Using the lightning package:
from lightning.pytorch.strategies import DDPStrategy
# or with the pytorch_lightning package:
# from pytorch_lightning.strategies import DDPStrategy

trainer = Trainer(..., strategy=DDPStrategy(process_group_backend="gloo"))
```

**Comment by andreemic** [comment]
> ```python
> from lightning.pytorch.strategies import DDPStrategy
> ```

what does setting the backend to "gloo" do? Is your proposed code setting the backend that is  described [here](https://pytorch.org/docs/stable/distributed.html)?

How will this impact training on GPUs?

**Comment by awaelchli** [comment]
Yes, it is the same as in PyTorch. Should be used only for debugging, not all communication collectives for GPU training are supported in gloo. If NCCL is set up correctly on your system (pytorch ships it automatically with the install), you should use this (trainer automatically selects it).

**Comment by PussyCat0700** [comment]
Guys I think I just found a solution for 4090 users: If you're running ddp with nccl on NVIDIA GeForce RTX 4090, lightning is likely to get stuck just in the way the issue described. In that case, simply add an env-variable in your terminal with:
`export NCCL_P2P_DISABLE=1`
With this env set, DDP should be running without a problem!

**Comment by PussyCat0700** [comment]
> @PussyCat0700 Oh my heavens, the predicament that had haunted me for an entire week has indeed been resolved. Throughout this past week, I was on the brink of despair while grappling with this conundrum, and I even felt as though my very hair might be shed in the process. Your assistance has been invaluable, and I extend my heartfelt gratitude. Might I inquire whether you possess the knowledge of how this particular environment variable has managed to provide a solution to this perplexity?

Glad to help out! Anyway, I was only advised by our hardware engineer so I don't know where you can find that kind of knowledge. Sorry about that!

**Comment by phrasenmaeher** [comment]
For those using SLURM, start you job using `srun python ...` instead of plain `python ...`
